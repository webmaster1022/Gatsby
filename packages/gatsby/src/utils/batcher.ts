export default class Batcher<Callback extends (...args: Array<any>) => any> {
  private queue: Array<Parameters<Callback>> = []
  private callbacks: Array<(...args: Parameters<Callback>) => void> = []
  private bulkCallbacks: Array<(args: Array<Parameters<Callback>>) => void> = []

  constructor(private threshold: number) {}

  /** Add a call to the batcher */
  add(...args: Parameters<Callback>): void {
    this.queue.push(args)

    if (this.queue.length >= this.threshold) {
      this.flush()
    }
  }

  /** Call all of our callbacks and clear out the queue */
  flush(): void {
    // call each callback for each item in the queue
    this.queue.forEach(args =>
      this.callbacks.forEach(callback => callback(...args))
    )

    // pass the entire queue to all bulk callbacks
    this.bulkCallbacks.forEach(callback => {
      callback(this.queue)
    })

    // clear out the queue
    this.queue = []
  }

  /** Sets up a callback for each batcher item */
  call(callback: (...args: Parameters<Callback>) => void): void {
    this.callbacks.push(callback)
  }

  /** Sets up a bulk callback that takes the entire queue */
  bulkCall(callback: (args: Array<Parameters<Callback>>) => void): void {
    this.bulkCallbacks.push(callback)
  }
}
