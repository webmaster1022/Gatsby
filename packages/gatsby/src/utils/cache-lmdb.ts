import { open, RootDatabase, Database, DatabaseOptions } from "lmdb"
import fs from "fs-extra"
import path from "path"

// Since the regular GatsbyCache saves to "caches" this should be "caches-lmdb"
const cacheDbFile =
  process.env.NODE_ENV === `test`
    ? `caches-lmdb-${
        // FORCE_TEST_DATABASE_ID will be set if this gets executed in worker context
        // when running jest tests. JEST_WORKER_ID will be set when this gets executed directly
        // in test context (jest will use jest-worker internally).
        process.env.FORCE_TEST_DATABASE_ID ?? process.env.JEST_WORKER_ID
      }`
    : `caches-lmdb`

export default class GatsbyCacheLmdb {
  private static store
  private db: Database | undefined
  private encoding: DatabaseOptions["encoding"]
  public readonly name: string
  // Needed for plugins that want to write data to the cache directory
  public readonly directory: string
  // TODO: remove `.cache` in v4. This is compat mode - cache-manager cache implementation
  // expose internal cache that gives access to `.del` function that wasn't available in public
  // cache interface (gatsby-plugin-sharp use it to clear no longer needed data)
  public readonly cache: GatsbyCacheLmdb

  constructor({
    name = `db`,
    encoding = `json`,
  }: {
    name: string
    encoding?: DatabaseOptions["encoding"]
  }) {
    this.name = name
    this.encoding = encoding
    this.directory = path.join(process.cwd(), `.cache/caches/${name}`)
    this.cache = this
  }

  init(): GatsbyCacheLmdb {
    fs.ensureDirSync(this.directory)
    return this
  }

  private static getStore(): RootDatabase {
    if (!GatsbyCacheLmdb.store) {
      GatsbyCacheLmdb.store = open({
        name: `root`,
        path: path.join(process.cwd(), `.cache/${cacheDbFile}`),
        compression: true,
        maxDbs: 200,
      })
    }
    return GatsbyCacheLmdb.store
  }

  private getDb(): Database {
    if (!this.db) {
      this.db = GatsbyCacheLmdb.getStore().openDB({
        name: this.name,
        encoding: this.encoding,
      })
    }
    return this.db
  }

  async get<T = unknown>(key): Promise<T | undefined> {
    return this.getDb().get(key)
  }

  async set<T>(key: string, value: T): Promise<T | undefined> {
    await this.getDb().put(key, value)
    return value
  }

  async del(key: string): Promise<void> {
    return this.getDb().remove(key) as unknown as Promise<void>
  }
}
