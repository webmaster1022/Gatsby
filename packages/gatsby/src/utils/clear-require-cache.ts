export function clearRequireCacheRecursively(
  module: string,
  ignoreNodeModules: boolean = true
): void {
  const resolvedModule = require.resolve(module)
  const cacheEntry = require.cache[resolvedModule]
  if (!cacheEntry) {
    return
  }
  if (ignoreNodeModules && resolvedModule.includes(`node_modules`)) {
    return
  }
  delete require.cache[resolvedModule]

  const parent = cacheEntry.parent
  if (parent) {
    const index = parent.children.findIndex(c => c.filename === resolvedModule)
    if (index !== -1) {
      parent.children.splice(index, 1)
    }
  }
  cacheEntry.children.forEach(module =>
    clearRequireCacheRecursively(module.filename, ignoreNodeModules)
  )
}
