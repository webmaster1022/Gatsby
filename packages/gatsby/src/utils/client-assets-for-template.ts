import * as path from "path"
import * as fs from "fs-extra"

// we want to force posix-style joins, so Windows doesn't produce backslashes for urls
const { join } = path.posix

export interface IScriptsAndStyles {
  scripts: Array<any>
  styles: Array<any>
  reversedStyles: Array<any>
  reversedScripts: Array<any>
}

interface IChunk {
  name: string
  rel: string
  content?: string
  shouldGenerateLink?: boolean
}

const inlineCssPromiseCache = new Map<string, Promise<string>>()

export async function readWebpackStats(publicDir: string): Promise<any> {
  const filePath = join(publicDir, `webpack.stats.json`)
  const rawPageData = await fs.readFile(filePath, `utf-8`)

  return JSON.parse(rawPageData)
}

export async function getScriptsAndStylesForTemplate(
  componentChunkName,
  webpackStats
): Promise<IScriptsAndStyles> {
  const uniqScripts = new Map<string, IChunk>()
  const uniqStyles = new Map<string, IChunk>()

  /**
   * Add script or style to correct bucket. Make sure those are unique (no duplicates) and that "preload" will win over any other "rel"
   */
  function handleAsset(
    name: string,
    rel: string,
    shouldGenerateLink: boolean = false
  ): void {
    let uniqueAssetsMap: Map<string, IChunk> | undefined

    // pick correct map depending on asset type
    if (name.endsWith(`.js`)) {
      uniqueAssetsMap = uniqScripts
    } else if (name.endsWith(`.css`)) {
      uniqueAssetsMap = uniqStyles
    }

    if (uniqueAssetsMap) {
      const existingAsset = uniqueAssetsMap.get(name)

      if (
        existingAsset &&
        rel === `preload` &&
        existingAsset.rel !== `preload`
      ) {
        // if we already track this asset, but it's not preload - make sure we make it preload
        // as it has higher priority
        existingAsset.rel = `preload`
      } else if (!existingAsset) {
        uniqueAssetsMap.set(name, { name, rel, shouldGenerateLink })
      }
    }
  }

  // Pick up scripts and styles that are used by a template using webpack.stats.json
  for (const chunkName of [`app`, componentChunkName]) {
    let assets = webpackStats.assetsByChunkName[chunkName]

    if (!assets) {
      continue
    }

    // Remove JS asset for templates
    if (chunkName !== `app`) {
      assets = assets.filter(asset => !asset.endsWith(`.js`))
    }

    for (const asset of assets) {
      if (asset === `/`) {
        continue
      }

      handleAsset(asset, `preload`)
    }

    // Handling for webpack magic comments, for example:
    // import(/* webpackChunkName: "<chunk_name>", webpackPrefetch: true */ `<path_to_module>`)
    // Shape of webpackStats.childAssetsByChunkName:
    // {
    //   childAssetsByChunkName: {
    //     <name_of_top_level_chunk>: {
    //       prefetch: [
    //         "<chunk_name>-<chunk_hash>.js",
    //       ]
    //     }
    //   }
    // }

    const childAssets = webpackStats.childAssetsByChunkName[chunkName]
    if (!childAssets) {
      continue
    }

    for (let [rel, assets] of Object.entries(childAssets)) {
      // Remove JS asset for templates(magic comments)
      if (chunkName !== `app`) {
        // @ts-ignore TS doesn't like that assets is not typed and especially that it doesn't know that it's Iterable
        assets = assets.filter(asset => !asset.endsWith(`.js`))
      }

      // @ts-ignore TS doesn't like that assets is not typed and especially that it doesn't know that it's Iterable
      for (const asset of assets) {
        // Use shouldGenerateLink to determines if  we should append link for magic comment asset(preload|prefetch) to head
        const shouldGenerateLink = chunkName == `app` ? true : false
        handleAsset(asset, rel, shouldGenerateLink)
      }
    }
  }

  // create scripts array, making sure "preload" scripts have priority
  const scripts: Array<IChunk> = []
  for (const scriptAsset of uniqScripts.values()) {
    if (scriptAsset.rel === `preload`) {
      // give priority to preload
      scripts.unshift(scriptAsset)
    } else {
      scripts.push(scriptAsset)
    }
  }

  // create styles array, making sure "preload" styles have priority and that we read .css content for non-prefetch "rel"s for inlining
  const styles: Array<IChunk> = []
  for (const styleAsset of uniqStyles.values()) {
    if (styleAsset.rel !== `prefetch`) {
      let getInlineCssPromise = inlineCssPromiseCache.get(styleAsset.name)
      if (!getInlineCssPromise) {
        getInlineCssPromise = fs.readFile(
          join(process.cwd(), `public`, styleAsset.name),
          `utf-8`
        )

        inlineCssPromiseCache.set(
          styleAsset.name,
          getInlineCssPromise as Promise<string>
        )
      }

      styleAsset.content = await getInlineCssPromise
    }

    if (styleAsset.rel === `preload`) {
      // give priority to preload
      styles.unshift(styleAsset)
    } else {
      styles.push(styleAsset)
    }
  }

  return {
    scripts,
    styles,
    reversedStyles: styles.slice(0).reverse(),
    reversedScripts: scripts.slice(0).reverse(),
  }
}

export function clearCache(): void {}
