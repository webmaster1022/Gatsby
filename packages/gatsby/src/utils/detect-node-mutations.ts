import reporter from "gatsby-cli/lib/reporter"
import { getNonGatsbyCodeFrameFormatted } from "./stack-trace-utils"
import type { IGatsbyNode } from "../redux/types"

const reported = new Set<string>()

const genericProxy = createProxyHandler()
const nodeInternalProxy = createProxyHandler({
  onGet(key, value) {
    if (key === `fieldOwners` || key === `content`) {
      // all allowed in here
      return value
    }
    return undefined
  },
  onSet(target, key, value) {
    if (key === `fieldOwners` || key === `content`) {
      target[key] = value
      return true
    }
    return undefined
  },
})

const nodeProxy = createProxyHandler({
  onGet(key, value) {
    if (key === `internal`) {
      return memoizedProxy(value, nodeInternalProxy)
    } else if (key === `fields` || key === `children`) {
      // all allowed in here
      return value
    }
    return undefined
  },
  onSet(target, key, value) {
    if (key === `fields` || key === `children`) {
      target[key] = value
      return true
    }
    return undefined
  },
})

/**
 * Every time we create proxy for object, we store it in WeakMap,
 * so that we reuse it for that object instead of creating new Proxy.
 * This also ensures reference equality: `memoizedProxy(obj) === memoizedProxy(obj)`.
 * If we didn't reuse already created proxy above comparison would return false.
 */
const referenceMap = new WeakMap<any, any>()
function memoizedProxy<T>(target: T, handler: ProxyHandler<any>): T {
  const alreadyWrapped = referenceMap.get(target)
  if (alreadyWrapped) {
    return alreadyWrapped
  } else {
    const wrapped = new Proxy(target, handler)
    referenceMap.set(target, wrapped)
    return wrapped
  }
}

function createProxyHandler({
  onGet,
  onSet,
}: {
  onGet?: (key: string | symbol, value: any) => any
  onSet?: (target: any, key: string | symbol, value: any) => boolean | undefined
} = {}): ProxyHandler<any> {
  function set(target, key, value): boolean {
    if (onSet) {
      const result = onSet(target, key, value)
      if (result !== undefined) {
        return result
      }
    }

    const error = new Error(`Stack trace:`)
    Error.captureStackTrace(error, set)

    if (error.stack && !reported.has(error.stack)) {
      reported.add(error.stack)
      const codeFrame = getNonGatsbyCodeFrameFormatted({
        stack: error.stack,
      })
      reporter.warn(
        `Node mutation detected\n\n${
          codeFrame ? `${codeFrame}\n\n` : ``
        }${error.stack.replace(/^Error:?\s*/, ``)}`
      )
    }
    return true
  }

  function get(target, key): any {
    const value = target[key]

    if (onGet) {
      const result = onGet(key, value)
      if (result !== undefined) {
        return result
      }
    }

    const fieldDescriptor = Object.getOwnPropertyDescriptor(target, key)
    if (fieldDescriptor && !fieldDescriptor.writable) {
      // this is to prevent errors like:
      // ```
      // TypeError: 'get' on proxy: property 'constants' is a read - only and
      // non - configurable data property on the proxy target but the proxy
      // did not return its actual value
      // (expected '[object Object]' but got '[object Object]')
      // ```
      return value
    }

    if (typeof value === `object` && value !== null) {
      return memoizedProxy(value, genericProxy)
    }

    return value
  }

  return {
    get,
    set,
  }
}

let shouldWrapNodesInProxies = !!process.env.GATSBY_DETECT_NODE_MUTATIONS
export function enableNodeMutationsDetection(): void {
  shouldWrapNodesInProxies = true

  reporter.warn(
    `Node mutation detection is enabled. Remember to disable it after you are finished with diagnostic as it will cause build performance degradation.`
  )
}

export function wrapNode<T extends IGatsbyNode | undefined>(node: T): T {
  if (node && shouldWrapNodesInProxies) {
    return memoizedProxy(node, nodeProxy)
  } else {
    return node
  }
}

export function wrapNodes<T extends Array<IGatsbyNode> | undefined>(
  nodes: T
): T {
  if (nodes && shouldWrapNodesInProxies && nodes.length > 0) {
    return nodes.map(node => memoizedProxy(node, nodeProxy)) as T
  } else {
    return nodes
  }
}
