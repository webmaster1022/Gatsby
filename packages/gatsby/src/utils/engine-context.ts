import { AsyncLocalStorage } from "async_hooks"

export interface IEngineContext {
  requestId: string
}

let asyncLocalStorage
function getAsyncLocalStorage(): AsyncLocalStorage<IEngineContext> {
  return asyncLocalStorage ?? (asyncLocalStorage = new AsyncLocalStorage())
}

export function getEngineContext(): IEngineContext | undefined {
  return getAsyncLocalStorage().getStore()
}

export function runWithEngineContext<T>(
  context: IEngineContext,
  fn: (...args: Array<any>) => T
): T {
  // @ts-ignore typings are incorrect, run() returns the result of fn()
  return getAsyncLocalStorage().run(context, fn)
}
