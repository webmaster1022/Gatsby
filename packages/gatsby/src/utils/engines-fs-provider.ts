// This module should be imported as first or one first modules in engines.
// It allows to provide alternative `fs` module by setting `global._fsWrapper`
// variable before importing engine. It will automatically fallback to regular
// `fs` if alternative `fs` is not provided.

/* eslint-disable @typescript-eslint/no-namespace */

declare global {
  namespace NodeJS {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    interface Global {
      _fsWrapper: typeof import("fs")
      _actualFsWrapper: typeof import("fs")
    }
  }
}

// eslint-disable-next-line @typescript-eslint/naming-convention
declare const __non_webpack_require__: typeof require

if (global._fsWrapper) {
  global._actualFsWrapper = global._fsWrapper
} else {
  // fs alternative not provided - falling back to regular fs
  global._actualFsWrapper = __non_webpack_require__(`fs`)
}

// hydrate webpack module cache (consume global, so it's not lazy)
require(`fs`)

// https://stackoverflow.com/a/59499895
export {}
