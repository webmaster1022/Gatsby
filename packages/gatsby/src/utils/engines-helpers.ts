import { emitter } from "../redux"
import { ICreatePageAction, ISetComponentFeatures } from "../redux/types"

export function shouldPrintEngineSnapshot(): boolean {
  return process.env.gatsby_executing_command === `build`
}

let generate = false
export function shouldGenerateEngines(): boolean {
  return process.env.gatsby_executing_command === `build` && generate
}

emitter.on(`CREATE_PAGE`, (action: ICreatePageAction) => {
  if (action.payload.mode && action.payload.mode !== `SSG`) generate = true
})
emitter.on(`SET_COMPONENT_FEATURES`, (action: ISetComponentFeatures) => {
  if (action.payload.serverData) generate = true
  if (action.payload.config) generate = true
})
