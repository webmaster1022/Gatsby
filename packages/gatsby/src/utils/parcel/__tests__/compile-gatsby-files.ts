import { Parcel } from "@parcel/core"
import {
  constructParcel,
  compileGatsbyFiles,
  gatsbyFileRegex,
  COMPILED_CACHE_DIR,
  PARCEL_CACHE_DIR,
} from "../compile-gatsby-files"
import { siteMetadata } from "./fixtures/utils/site-metadata"
import { moreDataConfig } from "./fixtures/utils/more-data-config"
import { readFile, remove, pathExists } from "fs-extra"

const dir = {
  js: `${__dirname}/fixtures/js`,
  ts: `${__dirname}/fixtures/ts`,
  tsOnlyInLocal: `${__dirname}/fixtures/ts-only-in-local-plugin`,
}

jest.setTimeout(15000)

jest.mock(`@parcel/core`, () => {
  const parcelCore = jest.requireActual(`@parcel/core`)

  const { Parcel: OriginalParcel } = parcelCore

  class MockedParcel extends OriginalParcel {
    constructor(options) {
      super(options)
      this.options = options // Expose for assertions
    }
  }

  return {
    ...parcelCore,
    Parcel: MockedParcel,
  }
})

interface IMockedParcel extends Parcel {
  options: unknown
}

let cwdToRestore
beforeAll(() => {
  cwdToRestore = process.cwd()
})

afterAll(() => {
  process.chdir(cwdToRestore)
})

describe(`gatsby file compilation`, () => {
  describe(`constructBundler`, () => {
    it(`should construct Parcel relative to passed directory`, () => {
      const { options } = constructParcel(dir.js) as IMockedParcel

      expect(options).toMatchSnapshot({
        entries: [
          `${dir.js}/${gatsbyFileRegex}`,
          `${dir.js}/plugins/**/${gatsbyFileRegex}`,
        ],
        targets: {
          root: {
            distDir: `${dir.js}/${COMPILED_CACHE_DIR}`,
          },
        },
        cacheDir: `${dir.js}/${PARCEL_CACHE_DIR}`,
      })
    })
  })

  describe(`compileGatsbyFiles`, () => {
    describe(`js files are not touched`, () => {
      beforeAll(async () => {
        process.chdir(dir.js)
        await remove(`${dir.js}/.cache`)
        await compileGatsbyFiles(dir.js)
      })

      it(`should not compile gatsby-config.js`, async () => {
        const isCompiled = await pathExists(
          `${dir.js}/.cache/compiled/gatsby-config.js`
        )
        expect(isCompiled).toEqual(false)
      })

      it(`should not compile gatsby-node.js`, async () => {
        const isCompiled = await pathExists(
          `${dir.js}/.cache/compiled/gatsby-node.js`
        )
        expect(isCompiled).toEqual(false)
      })
    })

    describe(`ts files are compiled`, () => {
      beforeAll(async () => {
        process.chdir(dir.ts)
        await remove(`${dir.ts}/.cache`)
        await compileGatsbyFiles(dir.ts)
      })

      it(`should compile gatsby-config.ts`, async () => {
        const compiledGatsbyConfig = await readFile(
          `${dir.ts}/.cache/compiled/gatsby-config.js`,
          `utf-8`
        )

        expect(compiledGatsbyConfig).toContain(siteMetadata.title)
        expect(compiledGatsbyConfig).toContain(siteMetadata.siteUrl)
        expect(compiledGatsbyConfig).toContain(moreDataConfig.options.name)
      })

      it(`should compile gatsby-node.ts`, async () => {
        const compiledGatsbyNode = await readFile(
          `${dir.ts}/.cache/compiled/gatsby-node.js`,
          `utf-8`
        )

        expect(compiledGatsbyNode).toContain(`I am working!`)
      })
    })

    describe(`ts only in local plugin files are compiled and outputted where expected`, () => {
      beforeAll(async () => {
        process.chdir(dir.tsOnlyInLocal)
        await remove(`${dir.tsOnlyInLocal}/.cache`)
        await compileGatsbyFiles(dir.tsOnlyInLocal)
      })

      it(`should compile gatsby-config.ts`, async () => {
        const compiledGatsbyConfig = await readFile(
          `${dir.tsOnlyInLocal}/.cache/compiled/plugins/gatsby-plugin-local/gatsby-config.js`,
          `utf-8`
        )

        expect(compiledGatsbyConfig).toContain(`gatsby-config is working`)
      })

      it(`should compile gatsby-node.ts`, async () => {
        const compiledGatsbyNode = await readFile(
          `${dir.tsOnlyInLocal}/.cache/compiled/plugins/gatsby-plugin-local/gatsby-node.js`,
          `utf-8`
        )

        expect(compiledGatsbyNode).toContain(`gatsby-node is working`)
      })
    })
  })
})
