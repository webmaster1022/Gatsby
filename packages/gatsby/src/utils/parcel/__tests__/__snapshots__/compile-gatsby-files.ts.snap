// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby file compilation constructBundler should construct Parcel relative to passed directory 1`] = `
Object {
  "cacheDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/.parcel-cache",
  "defaultConfig": "<PROJECT_ROOT>/packages/gatsby-parcel-config/lib/index.json",
  "entries": Array [
    "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/gatsby-+(node|config).ts",
    "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/plugins/**/gatsby-+(node|config).ts",
  ],
  "mode": "production",
  "targets": Object {
    "root": Object {
      "distDir": "<PROJECT_ROOT>/packages/gatsby/src/utils/parcel/__tests__/fixtures/js/.cache/compiled",
      "engines": Object {
        "node": ">= 14.15.0",
      },
      "includeNodeModules": false,
      "outputFormat": "commonjs",
      "sourceMap": false,
    },
  },
}
`;
